Create package rosidl_core_generators

jacobperron opened the pull request:
Move contents of rosidl_default_generators into the new package.
This package contains dependencies for code generation of ROS messages.
rosidl_default_generators now depends on the new package, plus message definitions required for Actions (namely action_msgs).

This allows users to avoid having to explictly depend on action_msgs.

Connected to

- https://github.com/ros2/rosidl_core/pull/1
- https://github.com/ros2/ros2/pull/1316
- https://github.com/ros2/rcl_interfaces/pull/144
- https://github.com/ros2/unique_identifier_msgs/pull/24

Downstream PRs removing the dependency:
- https://github.com/ros2/geometry2/pull/547
- https://github.com/ros2/example_interfaces/pull/16
- https://github.com/ros2/demos/pull/580
- https://github.com/ros2/ros2cli/pull/743

Documentation PRs:
- https://github.com/ros2/ros2_documentation/pull/2913
- TODO: Iron release note.

jacobperron commented:
Though this avoids a dependency cycle with action_msgs, I just realized this creates a cycle at the repository level. So, if this is an approved solution we should consider moving the new package to it's own repository.

jacobperron commented:
Note, I think we also require https://github.com/ros2/rosidl/pull/696 in order to avoid the dependency cycle.

jacobperron commented on rosidl_default_generators/package.xml:
```diff
@@ -9,32 +9,17 @@
   <buildtool_depend>ament_cmake</buildtool_depend>
+  
+  <!-- Code generation for Actions depends on action_msgs -->
+  <buildtool_export_depend>action_msgs</buildtool_export_depend>
```
Although this prevents downstream packages from having to declare a dependency on `action_msgs`, it now requires them to declare an export dependency on `rosidl_default_generators`. This is because packages building/running against a package that build an action may also need to link against (or import) `action_msgs`.

Maybe we could avoid having users export the `rosidl_default_generators` dependency if we avoid exposing `action_msgs` headers in public headers of generated interface libraries. I think there's only one place this is happening in the cpp generator. I'm not sure what the situation is in generated Python code though.

jacobperron replied:
#23 seems to resolve this issue by using rosidl_default_runtime to add an <exec_depend> on action_msgs.

ivanpauno replied:
IIRC, the cmake function to generate actions was showing an error if `action_msgs` was not listed as a dependency (somehow).
We should double check if that code still makes sense.

jacobperron replied:
Yes, the check is happening [here](https://github.com/ros2/rosidl/blob/b8381d955a111cdf8a52a0f1891cc55de5f19db1/rosidl_cmake/cmake/rosidl_generate_interfaces.cmake#L146-L151). IMO, it doesn't hurt to check. For instance, it's still possible someone will depend on `rosidl_core_generators` (or `rosidl_cmake`) directly. In both of those scenarios, the user still needs to depend on `action_msgs` explicitly if they have a custom action.

If anything, we could consider updating the error message to suggest they depend on `rosidl_default_generators` instead :shrug:

ivanpauno replied:
> IMO, it doesn't hurt to check

Yes, now that I see the code that was used to check, it seems fine.

> If anything, we could consider updating the error message to suggest they depend on rosidl_default_generators instead shrug

That would be okay

ivanpauno commented on rosidl_core_generators/CHANGELOG.rst:
```diff
@@ -0,0 +1,45 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
I think this should be generated when a new relase is done

jacobperron replied:
Maybe. My thinking was that this package was simply renamed, so it might be worth keeping the changelog here. However, considering we want to move this to a new repository I think starting a fresh changelog is okay too (since we probably will reset the version number :shrug:).

jacobperron replied:
After realizing the changelog doesn't contain much content, I'll probably remove this in the new repository

ivanpauno replied:
> Maybe. My thinking was that this package was simply renamed, so it might be worth keeping the changelog here.

Ah okay, that seems fine.
It's a metapackage though, so there's not much stuff of value in the changelogs.

If we keep the old logs, we should also start from the next version number.

ivanpauno commented on rosidl_default_generators/CMakeLists.txt:
```diff
@@ -5,13 +5,11 @@ project(rosidl_default_generators NONE)
 endif()
 
-ament_package(
-  CONFIG_EXTRAS "rosidl_default_generators-extras.cmake.in"
```
isn't this still needed for `rosidl_default_generators`?
I think that the idea is for it to work as a cmake "metapackage" as well.

I'm not sure if anybody was using the exported flags though (or using `rosidl_default_generators` in things like `ament_dependencies()`), but if someone would then the `-extras` file is still needed.

jacobperron replied:
I did a test, and it appears the cmake variables are set transitively. I.e. depending on rosidl_default_generators continues to set the variable `rosidl_default_generators_INCLUDE_DIRS`, for example.

ivanpauno replied:
that's pretty weird, did you clean your `install` folder first?
If not, the old config file will be there.

ivanpauno replied:
needs confirmation, otherwise the PR looks good to me

jacobperron replied:
I think it works because of the `ament_export_dependencies` macros in the CMakeLists.txt. I noticed that the variables were **not** being set when we `find_package(rosidl_default_runtime)`; adding `ament_export_dependencies` fixed it ([07dd659](https://github.com/ros2/rosidl_defaults/pull/22/commits/07dd659ceaeb8889ab7cb954675565e7f17621e9))

Tested with the following cmake code after a fresh build (removed the build and install directories):

```cmake
find_package(rosidl_default_generators REQUIRED)

message("default-definitions: ${rosidl_default_generators_DEFINITIONS}")
message("default-include_dirs: ${rosidl_default_generators_INCLUDE_DIRS}")
message("default-libraries: ${rosidl_default_generators_LIBRARIES}")
message("core-definitions: ${rosidl_core_generators_DEFINITIONS}")
message("core-include_dirs: ${rosidl_core_generators_INCLUDE_DIRS}")
message("core-libraries: ${rosidl_core_generators_LIBRARIES}")
message("")

find_package(rosidl_default_runtime REQUIRED)
message("default-runtime-definitions: ${rosidl_default_runtime_DEFINITIONS}")
message("default-runtime-include_dirs: ${rosidl_default_runtime_INCLUDE_DIRS}")
message("default-runtime-libraries: ${rosidl_default_runtime_LIBRARIES}")
message("core-runtime-definitions: ${rosidl_core_runtime_DEFINITIONS}")
message("core-runtime-include_dirs: ${rosidl_core_runtime_INCLUDE_DIRS}")
message("core-runtime-libraries: ${rosidl_core_runtime_LIBRARIES}")
```

All of the above variables are populated, except for the `*_DEFINITIONS` variables are never set (but I don't think that has anything to do with this change.

ivanpauno replied:
Nice, LGTM then

ivanpauno commented on rosidl_core_generators/rosidl_core_generators-extras.cmake.in:
```diff
@@ -1,5 +1,5 @@
+# rosidl_core_generators/rosidl_core_generators-extras.cmake.in
 
 find_package(ament_cmake_core REQUIRED)
 ament_index_get_resources(rosidl_typesupport_c_packages "rosidl_typesupport_c")
```
In L20-32, maybe the same could be done for `${_dep}_TARGETS`?

jacobperron replied:
I'm not sure I understand. What are you proposing be done?

jacobperron replied:
I mean, I think I understand what you are proposing, but I don't understand why. It seems unrelated to this PR.

ivanpauno replied:
> It seems unrelated to this PR.

Yes, it's a bit unrelated.
But now that we support targets we should be doing that in "cmake metapackages" as well, as that's the variable we're checking first when you use `ament_target_dependencies()` IIRC (and if not, we fallback to the old style variables)

jacobperron replied:
Let me propose that change separately from the package refactoring.

ivanpauno replied:
sounds good

ivanpauno reviewed (changes requested):
About the name of the new package, I don't have suggestions, but the distinction between "code" and "defaults" is confusing.

LGTM when the new package is moved to another repo (if not, it will create a circular dependency when bloom-releasing).

jacobperron commented:
@ivanpauno I've updated this PR, moving the new "core" packages their own repositories. I've updated this PRs description with an updated list of connected PRs.

ivanpauno reviewed (commented):
LGTM when the only open discussion is resolved (and CI is green)

jacobperron commented:
Full CI with all connected PRs:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=17151)](http://ci.ros2.org/job/ci_linux/17151/)
  * Same [test failure on nightly](https://ci.ros2.org/view/nightly/job/nightly_linux_repeated/2751/testReport/junit/(root)/rosbag2_transport/test_play_services__rmw_fastrtps_cpp_gtest_missing_result/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11696)](http://ci.ros2.org/job/ci_linux-aarch64/11696/)
* Windows [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_windows&build=17582)](https://ci.ros2.org/job/ci_windows/17582/)
